src/bindings/io/test/test_fd.c

```diff
@@ -0,0 +1,416 @@
+	binding->addTimer (binding, timerOp);
+
+	testSignalReadableCalled = 0;
+	testStop = stop;
+
+	start ();
+
+	succeed_if (testSignalReadableCalled, "callback was not called");
+
+	binding->removeFd (fdOp);
+	binding->removeTimer (timerOp);
+	binding->cleanup (binding);
+	elektraFree (fdOp);
+	elektraFree (timerOp);
+	close (fds[0]);
+	close (fds[1]);
+}
+
+static void testFdShouldUpdateEnabledControl (ElektraIoTimerOperation * timerOp ELEKTRA_UNUSED)
+{
+	switch (testUpdateEnabledStep)
+	{
+	case 0:
+		testUpdateEnabledFdOp->enabled = 1;
+		testUpdateEnabledBinding->updateFd (testUpdateEnabledFdOp);
+		break;
+	case 1:
+		yield_error ("timeout; test failed");
+		testStop ();
+		break;
+	}
+	testUpdateEnabledStep++;
+}
+
+static void testFdShouldUpdateEnabledProbe (ElektraIoFdOperation * fdOp ELEKTRA_UNUSED, int flags ELEKTRA_UNUSED)
+{
+	succeed_if (testUpdateEnabledStep != 0, "callback called before enabeld was updated");
+	testUpdateEnabledCalled = 1;
+	testStop ();
+}
+
+static void testFdShouldUpdateEnabled (ElektraIoTestSuiteCreateBinding createBinding, ElektraIoTestSuiteStart start,
+				       ElektraIoTestSuiteStop stop)
+{
+
+	int fds[2];
+	if (pipe (fds) == -1)
+	{
+		yield_error ("pipe() failed");
+		return;
+	}
+
+	ElektraIoFdOperation * fdOp = elektraIoTestSuiteUtilNewFdOperation ();
+	fdOp->fd = fds[1];
+	fdOp->flags = ELEKTRA_IO_WRITABLE;
+	fdOp->enabled = 0; // gets enabled by control timer
+	fdOp->callback = testFdShouldUpdateEnabledProbe;
+
+	ElektraIoTimerOperation * timerOp = elektraIoTestSuiteUtilNewTimerOperation ();
+	timerOp->interval = FD_CONTROL_INTERVAL;
+	timerOp->enabled = 1;
+	timerOp->callback = testFdShouldUpdateEnabledControl;
+
+	ElektraIoInterface * binding = createBinding ();
+	binding->addFd (binding, fdOp);
+	binding->addTimer (binding, timerOp);
+
+	testUpdateEnabledStep = 0;
+	testUpdateEnabledBinding = binding;
+	testUpdateEnabledFdOp = fdOp;
+	testUpdateEnabledCalled = 0;
+	testStop = stop;
+
+	start ();
+
+	succeed_if (testUpdateEnabledCalled, "callback was not called");
+
+	binding->removeFd (fdOp);
+	binding->removeTimer (timerOp);
+	binding->cleanup (binding);
+	elektraFree (fdOp);
+	elektraFree (timerOp);
+	close (fds[0]);
+	close (fds[1]);
+}
+
+static void testFdShouldUpdateFlagsControl (ElektraIoTimerOperation * timerOp ELEKTRA_UNUSED)
+{
+	switch (testUpdateFlagsStep)
+	{
+	case 0:
+		testUpdateFlagsFdOp->flags = ELEKTRA_IO_WRITABLE;
+		testUpdateFlagsBinding->updateFd (testUpdateFlagsFdOp);
+		break;
+	case 1:
+		yield_error ("timeout; test failed");
+		testStop ();
+		break;
+	}
+	testUpdateFlagsStep++;
+}
+
+static void testFdShouldUpdateFlagsProbe (ElektraIoFdOperation * fdOp ELEKTRA_UNUSED, int flags ELEKTRA_UNUSED)
+{
+	succeed_if (testUpdateFlagsStep != 0, "callback called before flags were updated");
+	testUpdateFlagsCalled = 1;
+	testStop ();
+}
+
+static void testFdShouldUpdateFlags (ElektraIoTestSuiteCreateBinding createBinding, ElektraIoTestSuiteStart start,
```

Comment:
Can you add a bit of explanations for the test cases? How is the pipe used?

Reply:
Ok - maybe a primer before the test cases? `pipe` is used to create simple (and POSIX compliant) file descriptors for the tests. Plus it's easy to use them to test the flags (e.g. wait for readability)